### PR TITLE
Update call to work with API key

### PR DESCRIPTION
**Pull request by alisaraa**
Hey all,
I know this has been tried before, but I read the feedback from the last submission. I have an API key to Fixerio (basic) and this worked for me. I updated the documentation.
Let me know if you have any questions, thanks,
Alisa

Hopefully this solves: https://github.com/singer-io/tap-fixerio/issues/3

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,6 +3,55 @@
...
 
+## How to use it
+
+`tap-fixerio` works to pull currency exchange rates from the Fixer API
+
+### Install and Run
+
+First, make sure Python 3 is installed on your system or follow these
+installation instructions for [Mac] or
+[Ubuntu].
```
Comment (timvisher) [inline README.md]: Line wrapping here is a little weird. I think `[Ubuntu].` should be on the previous line?

File: README.md [hunk 2]
```diff
@@ -3,6 +3,55 @@
 A [Singer](https://singer.io) Tap to extract currency exchange rate
 data from [fixer.io](http://fixer.io).
 
+## How to use it
+
+`tap-fixerio` works to pull currency exchange rates from the Fixer API
```
Comment (timvisher) [inline README.md]: This seems redundant with lines 3 and 4.

File: README.md [hunk 3]
```diff
@@ -3,6 +3,55 @@
...
+### Install and Run
+
+First, make sure Python 3 is installed on your system or follow these
+installation instructions for [Mac] or
+[Ubuntu].
+
+First, get an access key from [fixer.io](http://fixer.io). You need at least
+the basic plan to get exchange rates from a base currency (such as USD).
+
+Convert config.sample.json to config.json; fill out your parameters.
```
Comment (timvisher) [inline README.md]: Let's wrap config.sample.json and config.json in backticks.

File: README.md [hunk 4]
```diff
@@ -3,6 +3,55 @@
...
+It's recommended to use a virtualenv:
+
+```bash
+ python3 -m venv ~/.virtualenvs/tap-fixerio
+ source ~/.virtualenvs/tap-fixerio/bin/activate
+ pip install -U pip setuptools
+ pip install -e '.[dev]'
+```
+
+
```
Comment (timvisher) [inline README.md]: Little too much whitespace here.

File: README.md [hunk 5]
```diff
@@ -3,6 +3,55 @@
...
+These commands will install `tap-fixerio`  with pip,
+and then run it:
```
Comment (timvisher) [inline README.md]: These lines feel a little narrow, too? Hard wrapping should happen at 74.

File: README.md [hunk 6]
```diff
@@ -3,6 +3,55 @@
...
+
+```bash
+> pip install target-csv
```
Comment (timvisher) [inline README.md]: We should instead link to the setup instructions for `target-csv`, which should recommend setting up a dedicated `virtualenv` just for it to avoid version conflicts.

File: README.md [hunk 7]
```diff
@@ -3,6 +3,55 @@
...
+> tap-fixerio
```
Comment (timvisher) [inline README.md]: We shouldn't prefix these lines with `> ` so that they are easily copy/pastable.

Also, I believe this should be ```
~/.virtualenvs/tap-fixerio/bin/tap-fixerio --config something \
    | ~/.virtualenvs/target-csv/bin/target-csv --config something something
```

I would remove the example output below.

File: README.md [hunk 8]
```diff
@@ -3,6 +3,55 @@
...
+  INFO Replicating the latest exchange rate data from fixer.io
+  INFO Tap exiting normally
+```
+
+The data will be written to a file called `exchange_rate.csv` in your
+working directory.
+
+```bash
+› cat exchange_rate.csv
```
Comment (timvisher) [inline README.md]: This `> ` is useful (although I'd make it `$` as that's the default for bash `PS1` prompts.) as it shows that it's a console log. I would remove the `bash` language tag in this instance since it's just a plaintext log.

File: README.md (hunk 7 above)
Comment (alisaraa) [reply README.md]: Right, weirdly enough, the > were copied from https://github.com/singer-io/target-csv as was a lot of this. Do we want to be consistent with that?

File: README.md (hunk 8 above)
Comment (alisaraa) [reply README.md]: This was copied from here: https://github.com/singer-io/target-csv

I don't have strong feelings on it, but I thought being consistent was important?

File: README.md (hunk 2 above)
Comment (alisaraa) [reply README.md]: agreed, I felt weird even putting it. I was trying to be consistent with other READMEs but really my focus should have been on valuable written communication instead of copying and pasting.

File: README.md [hunk 9]
```diff
@@ -3,6 +3,49 @@
...
+ pip install -e '.[dev]'
+```
+
+Set up the `target-csv` virtual environment according to the instructions [here](https://github.com/singer-io/target-csv/blob/master/README.md).
+
+These commands will install `tap-fixerio`  with pip, and then run it:
+
+```bash
+  pip install target-csv
+  tap-fixerio
```
Comment (alisaraa) [inline README.md]: @timvisher  regarding -c config.json, I actually changed a line of the py file to make that file the default:
https://github.com/singer-io/tap-fixerio/pull/5/files#diff-1e7ded0f3322243ae4acd23372fd3a3aR85

Comment (timvisher) [reply README.md]: Ah that's interesting. Sorry I missed that.

I think I'm actually going to recommend that we remove that default. It's definitely off the beaten path in terms of behavior and I think it implies that there's a standard place to put the config file when there really isn't one yet.

FWIW, I would in the README put `~/singer.io/tap_fixerio_config.json` as the recommended place to put the config for now. I think I like the idea generally of having a standard place to put the configs but I'm not ready to make the call yet on where that should be. Opening an issue on the getting-started repo is probably a good way to start that conversation. :)

File: README.md [hunk 10]
```diff
@@ -3,6 +3,49 @@
...
+
+First, get an access key from [fixer.io](http://fixer.io). You need at least
+the basic plan to get exchange rates from a base currency (such as USD).
+
+Convert `config.sample.json` to `config.json`; fill out your parameters.
+
+It's recommended to use a virtualenv:
+
+```bash
+ python3 -m venv ~/.virtualenvs/tap-fixerio
```
Comment (timvisher) [inline README.md]: Let's delete all the leading whitespace here and in all other code blocks that we have.

File: README.md [hunk 11]
```diff
@@ -3,6 +3,49 @@
...
+
+### Install and Run
+
+First, make sure Python 3 is installed on your system or follow these
+installation instructions for [Mac] or [Ubuntu].
+
+First, get an access key from [fixer.io](http://fixer.io). You need at least
+the basic plan to get exchange rates from a base currency (such as USD).
+
+Convert `config.sample.json` to `config.json`; fill out your parameters.
```
Comment (timvisher) [inline README.md]: Here's where I would recommend `~/singer.io/tap_fixerio_config.json`.

File: README.md [hunk 12]
```diff
@@ -3,6 +3,49 @@
...
+
+It's recommended to use a virtualenv:
+
+```bash
+ python3 -m venv ~/.virtualenvs/tap-fixerio
+ source ~/.virtualenvs/tap-fixerio/bin/activate
+ pip install -U pip setuptools
+ pip install -e '.[dev]'
```
Comment (timvisher) [inline README.md]: Does `fixerio` have an `extras_require` block? The `.[dev]` is only needed if it does. Otherwise it can be `.`.

File: README.md [hunk 13]
```diff
@@ -3,6 +3,49 @@
...
+```
+
+Set up the `target-csv` virtual environment according to the instructions [here](https://github.com/singer-io/target-csv/blob/master/README.md).
```
Comment (timvisher) [inline README.md]: Can we make sure this line and all others are hard wrapped at 74 characters?

File: tap_fixerio.py [hunk 14]
```diff
@@ -45,21 +44,24 @@ def request(url, params):
...
     state = {'start_date': start_date}
     next_date = start_date
-    
+
     try:
         while True:
-            response = request(base_url + next_date, {'base': base})
+            response = request(base_url + next_date, {'base': base, 'access_key': access_key})
             payload = response.json()
+            print(payload)
```
Comment (timvisher) [inline tap_fixerio.py]: This should probably be deleted.

File: tap_fixerio.py [hunk 15]
```diff
@@ -45,21 +44,24 @@ def request(url, params):
...
 
             if datetime.strptime(next_date, DATE_FORMAT) > datetime.utcnow():
                 break
+            elif payload.get('error'):
+                break
```
Comment (timvisher) [inline tap_fixerio.py]: Should the be a `raise` or at least a `logger.error` or something?

File: tap_fixerio.py [hunk 16]
```diff
@@ -80,7 +82,7 @@ def main():
     parser = argparse.ArgumentParser()
 
     parser.add_argument(
-        '-c', '--config', help='Config file', required=False)
+        '-c', '--config', help='Config file', default='config.json', required=False)
```
Comment (timvisher) [inline tap_fixerio.py]: Let's remove this default (see my other response for details).

File: tap_fixerio.py [hunk 17]
```diff
@@ -100,8 +102,8 @@ def main():
 
     start_date = state.get('start_date',
                            config.get('start_date', datetime.utcnow().strftime(DATE_FORMAT)))
-
-    do_sync(config.get('base', 'USD'), start_date)
+    access_key = state.get('access_key', config.get('access_key'))
```
Comment (timvisher) [inline tap_fixerio.py]: Might be worth doing a bit of hygiene checking here since the config used to not require an `access_key` so that a nice error message is emitted.

File: config.sample.json [hunk 18]
```diff
@@ -1,4 +1,5 @@
 {
     "base": "AUD",
-    "start_date": "2017-02-02"
+    "start_date": "2018-11-12",
```
Comment (timvisher) [inline config.sample.json]: Would it make sense for this to be `<some recent date in YYYY-MM-DD format>`? That way we don't fall into the trap where the example config is constantly retreating into history? :)

File: README.md [hunk 19]
```diff
@@ -3,6 +3,49 @@
...
+
+These commands will install `tap-fixerio`  with pip, and then run it:
+
+```bash
+  pip install target-csv
```
Comment (timvisher) [inline README.md]: Shouldn't this be deleted because we're installing it in a separate environment now?

File: README.md (hunk 9 above)
Comment (timvisher) [inline README.md]: Doesn't this need to be that `~/.virtualenvs/tap-fixerio/bin/tap-fixerio …` command I mentioned earlier?

Comment (alisaraa) [reply README.md]: (to your first part) - that's fair. I actually added it because this README didn't explain how to use -c so I thought "ah, this must be a mistake, I shouldn't need to pass in the config file."
But if the README says to, the default shall be removed.

File: config.sample.json (hunk 18 above)
Comment (alisaraa) [reply config.sample.json]: yes! time did NOT begin in November 2018.

File: tap_fixerio.py (hunk 15 above)
Comment (alisaraa) [reply tap_fixerio.py]: Ah I see; I was following the convention of line 61, but line 61 is just quitting because it's done, not erring.

So this would be a good place regarding the access key hygiene. This is catching things like: API key invalid, API limit hit. So yeah, it should be logged and the other print removed. (line 59).

File: tap_fixerio.py (hunk 17 above)
Comment (alisaraa) [reply tap_fixerio.py]: See lines 107-110

File: README.md (hunk 7 above)
Comment (luandy64) [reply README.md]: (Taking over for @timvisher this week.)

Unfortunately there's not a ton of consistent styling between all the different documents so let's just DTRT thing here and not worry about the formatting elsewhere. :)

File: README.md (hunk 8 above)
Comment (luandy64) [reply README.md]: `$` is the correct value so it'd actually be better to open a PR to fix `target-csv` than it is to use `>` here.

File: tap_fixerio.py (hunk 15 above)
Comment (luandy64) [reply tap_fixerio.py]: I agree that "access key hygiene" should be maintained 👍 

If we encounter things like invalid API keys or hitting an API limit, then I think the `raise` is better choice here. Maybe something like:
```
elif payload.get('error'):
    raise APIError(payload['error'])
```

File: tap_fixerio.py [hunk 20]
```diff
@@ -100,8 +102,14 @@ def main():
 
     start_date = state.get('start_date',
                            config.get('start_date', datetime.utcnow().strftime(DATE_FORMAT)))
+    access_key = state.get('access_key', config.get('access_key'))
+
+    ## if access key is not supplied
+    if access_key is None:
+        logger.fatal("Please provide valid access key for Fixerio API")
+        sys.exit(-1)
```
Comment (luandy64) [inline tap_fixerio.py]: Assuming we already have some `class APIError(Exception)`, this is a great spot to also use it and we can even recycle the `"Please provide valid access key for Fixerio API"`.

File: README.md [hunk 21]
```diff
@@ -3,6 +3,48 @@
...
 data from [fixer.io](http://fixer.io).
 
+## How to use it
+
+### Install and Run
+
+First, make sure Python 3 is installed on your system or follow these
+installation instructions for [Mac] or [Ubuntu].
+
+First, get an access key from [fixer.io](http://fixer.io). You need at least
```
[the remark this answers is not in the document]
Comment (alisaraa) [reply README.md]: First: access key 
Then: take over the world.

File: README.md [hunk 22]
```diff
@@ -3,6 +3,48 @@
...
+These commands will install `tap-fixerio`  with pip, and then run it:
+
+```bash
+~/.virtualenvs/tap-fixerio/bin/tap-fixerio --config ~/singer.io/tap_fixerio_config.json
+```
+
+The data will be written to a file called `exchange_rate.csv` in your
+working directory.
+
+$ cat exchange_rate.csv
```
[the remark this answers is not in the document]
Comment (alisaraa) [reply README.md]: I believe @timvisher  suggested I remove them since it's not code- it's plain text.

File: tap_fixerio.py (hunk 20 above)
Comment (alisaraa) [reply tap_fixerio.py]: Where would I find if we have a class APIError implemented already?

Comment (luandy64) [reply tap_fixerio.py]: Oops, I was saying that assuming we made it to fix the change around line 63.

File: README.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (luandy64) [reply README.md]: I believe he was just talking about the `>` and the `bash`. The backticks should have been left untouched so that the markdown file renders the log in a different font.

File: tap_fixerio.py (hunk 15 above)
Comment (alisaraa) [reply tap_fixerio.py]: Should I raise instead of log?

Comment (luandy64) [reply tap_fixerio.py]: I think so. If we just `log`, then I think the tap can still run. 

I would rather it fail if it hits an error.

File: tap_fixerio.py (hunk 20 above)
Comment (dmosorast) [reply tap_fixerio.py]: This should also be raising an exception, like the change implemented in [this commit](https://github.com/singer-io/tap-fixerio/pull/5/commits/deef04d8e21b397cafeb9df63daf19a20ba872a3).

There are a few ways to get this behavior, but the most standard would be using the [singer-python argparser](https://github.com/singer-io/singer-python/blob/master/singer/utils.py#L126), with a list of required config keys. It will raise exceptions on its own when it parses. ([Example](https://github.com/singer-io/tap-zendesk/blob/master/tap_zendesk/__init__.py#L161))

File: README.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (dmosorast) [reply README.md]: Looks like this one is still outstanding.

**Comment by dmosorast** [comment]
Hi @alisaraa I just went through and cleaned up the conversations based on your recent commits. Looks like there are 2 still outstanding that I left comments on. It was a bit hard to follow, so if they're actually completed in the most recent version, please ignore 😄 

Making those changes should give me enough to merge this.

**Comment by KAllan357** [comment]
Awesome thanks :shipit:
